src/arrays.ts

```diff
@@ -158,6 +158,18 @@ export function join(...arrays: any[][]) {
   return arrays.reduce((a, x) => a.concat(x), []);
 }
 
+/** 
+ * Allows filtering and mapping over an array in a single step.
+ * Returns a list of values from `items` for which `predicate` 
+ * returns a non-null value.
+ */
+function filterMap<InputType, ReturnType>(
+    items: InputType[],
+    predicate: (item: InputType) => ReturnType | null
+): ReturnType[] {
+  const processedItems = items.map(item => predicate(item));
+  return processedItems.filter(processedItem => processedItem != null) as ReturnType[];
+}
```

Comment:
I'm not sure where you're using this (usually doing `.map().filter()` explicitly shouldn't be too much work), but for performance, I would make this a generator function:

``` ts
export function* filterMap<S, T>(items: Iterable<S>, predicate: (item: S) => T|undefined): Iterable<T> {
  for (const i of items) {
    const j = predicate(i);
    if (j !== undefined) yield j;
  }
}
```

This means that this function could be used for `Set`s and `Map`s, not just `Array`s. And since you can `for ... of` over `Iterable`s, the usage should be almost the same.

Reply:
Here is some background: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*

Reply:
It allows for a concise expression of cases like the following:
`const fibs = filterMap(numbers, fibonacci);`,
where `fibonacci(number)` returns `undefined` when `number` is less than 0.

The equivalent otherwise would be something like: `const fibs = numbers.map(fibonacci).filter(result => result != undefined);`